template-manager-cmtemplate/src/main/java/com/sequenceiq/cloudbreak/cmtemplate/CmTemplateProcessor.java

```diff
@@ -291,6 +311,52 @@ public String getStackVersion() {
         return cmTemplate.getCdhVersion();
     }
 
+    @Override
+    public Map<String, Map<String, ServiceAttributes>> getHostGroupBasedServiceAttributes() {
+        // The only attributes required at the moment relate to the YARN NodeManager.
+        // This functionality can be moved to an interface / separated if additional components
+        // need this functionality.
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponents = getNonGatewayServicesByHostGroup();
+        Map<String, Set<ServiceComponent>> hgToNonGwServiceComponentsWithYarnNMs = new HashMap<>();
+        for (Entry<String, Set<ServiceComponent>> entry : hgToNonGwServiceComponents.entrySet()) {
+            for (ServiceComponent sc : entry.getValue()) {
+                if (sc.getService().equalsIgnoreCase(YarnRoles.YARN) && sc.getComponent()
+                        .equalsIgnoreCase(YarnRoles.NODEMANAGER)) {
+                    hgToNonGwServiceComponentsWithYarnNMs.put(entry.getKey(), entry.getValue());
+                    break;
+                }
+            }
+        }
```

Comment:
it would be nicer with stream:
```
Map<String, Set<ServiceComponent>> hgToNonGwServiceComponentsWithYarnNMs = hgToNonGwServiceComponents.entrySet().stream()
                .filter(this::isYarnNodemanager)
                .collect(toMap(Entry::getKey, Entry::getValue));

private boolean isYarnNodemanager(Entry<String, Set<ServiceComponent>> entry) {
        return entry.getValue().stream().anyMatch(sc -> YarnRoles.YARN.equalsIgnoreCase(sc.getService())
                && YarnRoles.NODEMANAGER.equalsIgnoreCase(sc.getComponent()));
    }
```

Reply:
I don't like using Java streams (not as easy to read). If that's the pattern that is generally used, I can switch it over. (Applies to other comments related to streams)